### PR TITLE
Mheavin snow 640361 php keypair

### PR DESCRIPTION
**Pull request by sfc-gh-mheavin**
I added a subsection for Using Key Pair Authentication.

### CODE REVIEW

File: README.rst
Comment (sfc-gh-dyoshinaga) [inline README.rst:261]: Do you have a separate PR to add the PHP PDO driver to https://docs.snowflake.com/en/user-guide/key-pair-auth.html#supported-snowflake-clients ?

Comment (sfc-gh-mheavin) [reply README.rst:261]: No, but I will create one now. Or should I just add it to this PR?

File: README.rst
Comment (sfc-gh-dyoshinaga) [inline README.rst:278]: To make this easier to see that this in not the user's password, replace "<password>" with "<private_key_passphrase>"

Comment (sfc-gh-mheavin) [reply README.rst:278]: Done

File: README.rst
Comment (sfc-gh-mheavin) [reply README.rst:261]: SNOW-649074

File: README.rst
Comment (sfc-gh-dyoshinaga) [inline README.rst:286]: For consistency with https://docs.snowflake.com/en/user-guide/key-pair-auth.html#step-1-generate-the-private-key and the other docs (e.g. https://docs.snowflake.com/en/user-guide/jdbc-parameters.html#private-key-file-pwd ), we should use the term "passphrase", not "passcode".

We should also use the term "decrypt", rather than "decode".

File: README.rst
Comment (sfc-gh-dyoshinaga) [inline README.rst:285]: Should filenames be in code font, rather than in bold?

File: README.rst
Comment (sfc-gh-dyoshinaga) [inline README.rst:286]: Also change "<password>" to "<private_key_passphrase>" here.

Comment (sfc-gh-mheavin) [reply README.rst:286]: Done

Comment (sfc-gh-dyoshinaga) [inline README.rst]: As shown in https://docs.snowflake.com/en/user-guide/key-pair-auth.html#step-1-generate-the-private-key , we make a distinction between support for encrypted private keys and support for unencrypted private keys.

Do we also support unencrypted private keys? (in which case you'd omit the passphrase for the private key) Or do we only support encrypted private keys?

Comment (sfc-gh-mheavin) [reply README.rst]: We support both.

Adding verbiage to indicate it is optional for unencrypted keys

File: README.rst
Comment (sfc-gh-dyoshinaga) [inline README.rst:278]: I think this syntax is not correct.

1. This example shows each DSN property (AUTHENTICATOR, PRIV_KEY_FILE, PRIV_KEY_FILE_PWD) being passed as a separate argument.
2. The example doesn't show the username being passed in as a separate argument.

I think that:

1. Each of the new DSN properties needs to be part of the first argument and delimited by semicolons.  The [test case](https://github.com/snowflakedb/pdo_snowflake/pull/289/files#diff-0272c80a6ed5c366c86659fb7177d2788789fd918781941e3e55f9b57380f928) seems to indicate that this is how you should specify this.
2. In the test case also shows the username being passed in as a separate argument. I think in all of our drivers, when we authenticate using a private key, we still need to specify a username (e.g. the [Python Connector](https://docs.snowflake.com/en/user-guide/python-connector-example.html#using-key-pair-authentication-key-pair-rotation) example).

    The test cases also pass in the password argument, which makes me wonder if the argument is required and if it should be an empty string if you're using a private key (rather than a password) for authentication.

Comment (sfc-gh-mheavin) [reply README.rst:278]: Dev approved it, but I will run it by them again

File: README.rst
Comment (sfc-gh-mheavin) [reply README.rst]: Fixed

File: README.rst
Comment (sfc-gh-dyoshinaga) [reply README.rst]: I think we're providing our own implementation of this interface, and the constructor doesn't let you pass in arguments specifically for the authenticator, private key file, and private key file passphrase:

https://www.php.net/manual/en/pdo.connections.php
https://www.php.net/manual/en/pdo.construct.php

Any other information that needs to be passed in needs to go into the argument for the DSN or possibly the options array.

**Comment by sfc-gh-dyoshinaga** [comment]
One more thing: At this point in time, it looks like the PR to add this feature ( https://github.com/snowflakedb/pdo_snowflake/pull/289 ) hasn't been merged in yet.

We should wait until that PR gets merged in before merging in the doc changes.

File: README.rst
Comment (sfc-gh-dyoshinaga) [inline README.rst:286]: "Specifies to" => "Specifies that you want to"

Comment (sfc-gh-mheavin) [reply README.rst:286]: Done

File: README.rst
Comment (sfc-gh-mheavin) [reply README.rst]: Reaching out to dev again.

Comment (sfc-gh-dyoshinaga) [reply README.rst]: I think the example is still wrong. The example passes in five arguments:

- The string containing the DSN.
- The string containing the username (but in this odd "username=<username>" format, which doesn't seem to be correct).
- The string containing the authenticator parameter.
- The string containing the private key file parameter.
- The string containing the private key passphrase parameter.

We point to the PDO documentation, but their docs don't have anything about a constructor that passes in five string arguments:

https://www.php.net/manual/en/pdo.connections.php
https://www.php.net/manual/en/pdo.construct.php

I think the example is supposed to look like this:

```
$dbh = new PDO("snowflake:account=<account_name>; authenticator=snowflake_jwt;priv_key_file=<path_to_private_key_file.;priv_key_file_pwd=<private_key_passphrase>", "<user>", "<password>");
```

If the "<password>" argument is required, I'm assuming that it should be a blank string if you're using the key pair for authentication.

Comment (sfc-gh-mheavin) [reply README.rst]: I got a new example from dev (he apologized for not checking it closely) and am integrating it.

File: README.rst
Comment (sfc-gh-dyoshinaga) [inline README.rst:288]: Maybe start with the condition:

```
If you are using an unencrypted private key file, omit this parameter.
```

Comment (sfc-gh-mheavin) [reply README.rst]: Done

File: README.rst
Comment (sfc-gh-dyoshinaga) [inline README.rst:276]: `<username>` should be `"<username>"`, as it is on line 244.

Comment (sfc-gh-mheavin) [reply README.rst:276]: Done

File: README.rst
Comment (sfc-gh-dyoshinaga) [inline README.rst:284]: "Specifies" should be "specifies". (All of these items are sentences that start with the parameter name and placeholder. i.e. "specifies" is not the first word in the sentence.)

File: README.rst
Comment (sfc-gh-dyoshinaga) [inline README.rst:269]: "key par" should be "key pair"

File: README.rst
Comment (sfc-gh-mheavin) [reply README.rst:284]: Done

File: README.rst
Comment (sfc-gh-dyoshinaga) [inline README.rst:280]: If the goal is to have all the items use "Specified", the first item doesn't do that.

**Review by sfc-gh-dyoshinaga — APPROVED** [review]
Thanks for making these changes and fixing the incorrect example!

File: README.rst
Comment (sfc-gh-mheavin) [reply README.rst]: Done

**Pull request merged**
